Better support of empty directories

vadmeste opened the pull request:
## Description
Support of empty directories

## Motivation and Context
Better support of legacy applications following posix behavior when working with S3

## How Has This Been Tested?
Manually

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

vadmeste committed:
More enhanced support of empty directories

Better support of HEAD and listing of zero sized objects with trailing
slash (a.k.a empty directory). For that, isLeafDir function is added
to indicate if the specified object is an empty directory or not. Each
backend (xl, fs) has the responsibility to store that information.
Currently in both of XL & FS, a empty directory is represented by
an empty directory in the backend.

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -890,6 +894,15 @@ func (fs *FSObjects) listDirFactory(isLeaf isLeafFunc) listDirFunc {
 	return listDir
 }
 
+// isObjectDir
```
elaborate on the comment here @vadmeste

vadmeste replied:
> elaborate on the comment here @vadmeste

Done

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -1019,8 +1032,14 @@ func (fs *FSObjects) ListObjects(ctx context.Context, bucket, prefix, marker, de
 			// object string does not end with "/".
 			return !hasSuffix(object, slashSeparator)
 		}
+		isLeafDir := func(bucket, object string) bool {
```
Provide an explanation of what this does?

vadmeste replied:
> Provide an explanation of what this does?

Done

harshavardhana commented on cmd/tree-walk.go:
```diff
@@ -98,6 +98,8 @@ type listDirFunc func(bucket, prefixDir, prefixEntry string) (entries []string,
 // 4. XL backend multipart listing - isLeaf is true if the entry is a directory and contains uploads.json
 type isLeafFunc func(string, string) bool
 
+type isLeafDirFunc func(string, string) bool
```
you need to explain this function just like `isLeafFunc`

vadmeste replied:
> you need to explain this function just like isLeafFunc

Done

harshavardhana commented on cmd/tree-walk_test.go:
```diff
@@ -479,6 +505,15 @@ func TestSortedness(t *testing.T) {
 	}
+
+	isLeafDir := func(volume, prefix string) bool {
+		entries, listErr := disk1.ListDir(volume, prefix)
```
List dir would list all entries perhaps we should list only 1 entry?

vadmeste replied:
> List dir would list all entries perhaps we should list only 1 entry?

Done

harshavardhana commented on cmd/xl-v1-common.go:
```diff
@@ -76,6 +76,32 @@ func (xl xlObjects) isObject(bucket, prefix string) (ok bool) {
 	return false
 }
 
+// isObjectDir
```
expand the comment and elaborate..

vadmeste replied:
> expand the comment and elaborate..

Done

vadmeste committed:
Accelerate testing if a directory is empty in isLeafDir()

isLeafDir() checks if the given path is an empty directory or not,
since dir listing is costly if the latter contains too many objects,
readDirN() is added in this PR to list only N number of entries.
In isLeadDir(), we will only list one entry to check if a directory
is empty or not.

poornas reviewed (commented):
@vadmeste, travis build is failing. Could you give more context on what issue this PR is addressing

poornas commented on cmd/posix-list-dir_test.go:
```diff
@@ -201,3 +201,40 @@ func TestReadDir(t *testing.T) {
+		{2, 1, 1},
+		{10, 9, 9},
+		{10, 10, 10},
+		{10, 11, 10},
```
@vadmeste , could you add a test case for n = -1?

vadmeste replied:
> @vadmeste , could you add a test case for n = -1?

Fixed

poornas reviewed (approved):
LGTM.
